cpp/src/arrow/compute/memory_resources.cc

```diff
@@ -0,0 +1,307 @@
+  ARROW_ASSIGN_OR_RAISE(auto input, filesystem->OpenInputFile(file_path));
+  ARROW_ASSIGN_OR_RAISE(auto reader, arrow::ipc::feather::Reader::Open(input));
+  std::shared_ptr<Table> table;
+  ARROW_RETURN_NOT_OK(reader->Read(&table));
+  TableBatchReader batch_iter(*table);
+  ARROW_ASSIGN_OR_RAISE(auto batch, batch_iter.Next());
+  return batch;
+}
+
+class DiskDataHolder : public DataHolder {
+ public:
+  DiskDataHolder(const std::shared_ptr<RecordBatch>& record_batch,
+                 MemoryPool* memory_pool)
+      : DataHolder(MemoryLevel::kDiskLevel), memory_pool_(memory_pool) {
+    std::string root_path;
+    std::string file_name = "data-holder-temp-" + RandomString(64) + ".feather";
+
+    filesystem_ =
+        arrow::fs::FileSystemFromUri(cache_storage_root_path, &root_path).ValueOrDie();
+
+    file_path_ = root_path + file_name;
+    status_ = StoreRecordBatch(record_batch, filesystem_, file_path_);
+  }
+
+  Result<ExecBatch> Get() override {
+    ARROW_RETURN_NOT_OK(status_);
+    ARROW_ASSIGN_OR_RAISE(auto record_batch, RecoverRecordBatch(filesystem_, file_path_));
+    return ExecBatch(*record_batch);
+  }
+
+ private:
+  std::string file_path_;
+  Status status_;
+  MemoryPool* memory_pool_;
+  std::shared_ptr<arrow::fs::FileSystem> filesystem_;
+  const std::string cache_storage_root_path = "file:///tmp/";
+};
+
+class MemoryResources::MemoryResourcesImpl {
+ public:
+  Status AddMemoryResource(std::unique_ptr<MemoryResource> resource) {
+    std::lock_guard<std::mutex> mutation_guard(lock_);
+    auto level = resource->memory_level();
+    auto it = stats_.find(level);
+    if (it != stats_.end()) {
+      return Status::KeyError("Already have a resource type registered with name: ",
+                              resource->ToString());
+    }
+    stats_[level] = std::move(resource);
+    return Status::OK();
+  }
+
+  size_t size() const { return stats_.size(); }
+
+  Result<int64_t> memory_limit(MemoryLevel level) const {
+    auto it = stats_.find(level);
+    if (it == stats_.end()) {
+      return Status::KeyError("No memory resource registered with level: ",
+                              MemoryLevelName(level));
+    }
+    return it->second->memory_limit();
+  }
+
+  Result<int64_t> memory_used(MemoryLevel level) const {
+    auto it = stats_.find(level);
+    if (it == stats_.end()) {
+      return Status::KeyError("No memory resource registered with level: ",
+                              MemoryLevelName(level));
+    }
+    return it->second->memory_used();
+  }
+
+  Result<MemoryResource*> memory_resource(MemoryLevel level) const {
+    auto it = stats_.find(level);
+    if (it == stats_.end()) {
+      return Status::KeyError("No memory resource registered with level: ",
+                              MemoryLevelName(level));
+    }
+    return it->second.get();
+  }
+
+ private:
+  std::mutex lock_;
+
+  std::unordered_map<MemoryLevel, std::unique_ptr<MemoryResource>> stats_;
```

Comment:
Given that `MemoryLevel` is an enum could this just be a fixed array?

Reply:
Sure